File: 740. Delete and Earn/740. Delete and Earn bottom-up.java
class Solution {
    public int deleteAndEarn(int[] nums) {
        int[] count = new int[10001];
        for (int x : nums)
            count[x]++;

        int prev = -1, avoid = 0, using = 0;

        for (int i = 0; i <= 10000; i++)
            if (count[i] > 0) {
                int m = Math.max(avoid, using);
                if (i - 1 == prev) {
                    using = i * count[i] + avoid;
                    avoid = m;
                } else {
                    using = i * count[i] + m;
                    avoid = m;
                }
                prev = i;
            }
        return Math.max(avoid, using);
    }
}

File: 740. Delete and Earn/740. Delete and Earn bottom-up.py
from typing import List


class Solution:
    def deleteAndEarn(self, nums: List[int]) -> int:
        nums_counter = Counter(nums)
        prev = -1
        avoid = using = 0
        for i in sorted(nums_counter):
            if i - 1 != prev:
                avoid, using = max(avoid, using), i * \
                    nums_counter[i] + max(avoid, using)
            else:
                avoid, using = max(avoid, using), i*nums_counter[i]+avoid
            prev = i
        return max(avoid, using)
